### ISSUE TITLE
Multiple jobs in parallel in one node process

### ISSUE THREAD

**Issue by janober** [veny/GearmaNode#28]
Hello,

just started to play around with gearman and wanted to allow each node-process to handle multiple jobs at the same time. Wanted to avoid the overhead of multiple node processes.

So wanted to see if simply calling:
```
worker.addFunction('whatever', ...)
```
multiple times in the same node-app does the trick.

The simple test I did worked fine as long as I did start the client first and then the single worker with the 4x "addFunction". It then did complete the jobs roughly 4x faster. Can then see that it starts 4 jobs directly.
However when I first start the worker and then the client it does not seem to work. It then processes one job after another with exactly the same speed as a worker with just one "addFunction".

So was wondering if somebody knows what is going on there or what the proper approach should be to achieve my goal.

Thanks!

**Comment by janober** [comment]
Ups just realized that I should probably also add
```
var worker = gearmanode.worker(...);
```
into the creation loop. As soon as I do that, it works fine no matter what I start first. However would still be interested in knowing if that is ok to do or if that could cause any problems anybody can think of.

Thanks!

**Comment by veny** [comment]
hi Jan,
node.js is a single thread technology. it means one node.js process provides for you only one thread that can process data received from gearman server.
with other words, it does not matter how many workers you register in one node.js process, only one can work in given time.

**Comment by janober** [comment]
Ah yes I am aware of that. And you are right it would probably be totally wrong for the most use cases. However my workers are parsing websites and the most of the time they are just sitting there and doing nothing because they are waiting for the response. And then does a lot of stuff happen in PhantomJS which is another task anyway.

So ignoring the single thread thing are you aware of any issues that my above approach could cause with Gearman/GearmaNode?

**Comment by veny** [comment]
if your worker waits for a HTTP response of a website to be parsed, it is marked on gearman server as a busy worker and the server does not assign a new task.
but, you can start more node.js processes. this is usual approach for parallelization. not on thread level, but on process level.

**Comment by janober** [comment]
Ah yes that would be the easiest and obvious thing to do. Just wanted to avoid the overhead of multiple node processes (did honestly not really test/check how much it it really is per process). In my case I run it currently on very very small virtual machines (most time just 1 CPU with 0.5-1 GB RAM). So no matter if I start one process or multiple ones, the one CPU is the limit.

By running it all in the same process I hope to save RAM and probably also a little little bit CPU.

Sure will not always be like that and later will use multiple processes (when I use bigger machines), however right now while I just develop and test, it seems like the best/cheapest solution. That is why I was wondering if there are any problems to expect. However until now it seems to work fine and does exactly what expected.

**Comment by janober** [comment]
Simply assume everything will be fine because that seems to be the case. 

So I close it.
